control-plane-operator/hostedclusterconfigoperator/controllers/globalps/globalps.go

```diff
@@ -0,0 +1,392 @@
+package globalps
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/openshift/hypershift/control-plane-operator/hostedclusterconfigoperator/controllers/resources/manifests"
+	"github.com/openshift/hypershift/support/thirdparty/kubernetes/pkg/credentialprovider"
+	"github.com/openshift/hypershift/support/upsert"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	crreconcile "sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	ControllerName = "globalps"
+)
+
+type Reconciler struct {
+	cpClient               crclient.Client
+	kubeSystemSecretClient crclient.Client
+	hcUncachedClient       crclient.Client
+	hcpNamespace           string
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req crreconcile.Request) (crreconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("reconciling global pull secret")
+
+	// Reconcile GlobalPullSecret
+	hccoImage := os.Getenv("HOSTED_CLUSTER_CONFIG_OPERATOR_IMAGE")
+	if hccoImage == "" {
+		return ctrl.Result{}, fmt.Errorf("HOSTED_CLUSTER_CONFIG_OPERATOR_IMAGE is not set")
+	}
+
+	if err := r.reconcileGlobalPullSecret(ctx, r.hcpNamespace, hccoImage); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile global pull secret: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcileGlobalPullSecret reconciles the original pull secret given by HCP and merges it with a new pull secret provided by the user.
+// The new pull secret is only stored in the DataPlane side so, it's not exposed in the API. It lives in the kube-system namespace of the DataPlane.
+// If that PS exists, the HCCO deploys a DaemonSet which mounts the whole Root FS of the node, and merges the new PS with the original one.
+// If the PS doesn't exist, the HCCO doesn't do anything.
+func (r *Reconciler) reconcileGlobalPullSecret(ctx context.Context, hcpNamespace string, cpoImage string) error {
+	var (
+		userProvidedPullSecretBytes []byte
+		originalPullSecretBytes     []byte
+		globalPullSecretBytes       []byte
+		err                         error
+	)
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("reconciling global pull secret")
+
+	// Get the user provided pull secret
+	exists, additionalPullSecret, err := AdditionalPullSecretExists(ctx, r.kubeSystemSecretClient)
+	if err != nil {
+		return fmt.Errorf("failed to check if user provided pull secret exists: %w", err)
+	}
+
+	// Reconcile the RBAC for the Global Pull Secret
+	if err := ReconcileGlobalPullSecretRBAC(ctx, r.hcUncachedClient, r.kubeSystemSecretClient, r.CreateOrUpdate, "kube-system", "openshift-config"); err != nil {
```

Comment:
why reconcile the RBAC here before checking if the additionalPullSecret exist? This should be moved down just before where "ReconcileDaemonSet" is called.

Reply:
Yeah, makes sense. It has more sense before when the RBAC and DS were deleted. Thanks!